Update JS perf article and others (#27848)
mdn/content: files/en-us/learn/performance/javascript/index.md

```js
function loop() {
  // Clear the camvas before drawing the next frame of the animation
  ctx.fillStyle = "rgba(0, 0, 0, 0.25)";
  ctx.fillRect(0, 0, width, height);

  // Draw objects on the canvas and update their positioning data
  // ready for the next frame
  for (const ball of balls) {
    ball.draw();
    ball.update();
  }

  // Call requestAnimationFrame to run the loop() function again
  // at the right time to keep the animation smooth
  requestAnimationFrame(loop);
}

// Call the loop() function once to set the animation running
loop();
```

You can find a nice introduction to canvas animations at [Drawing graphics > Animations](/en-US/docs/Learn/JavaScript/Client-side_web_APIs/Drawing_graphics#animations), and a more in-depth example at [Object building practice](/en-US/docs/Learn/JavaScript/Objects/Object_building_practice). You can also find a full set of canvas tutorials at [Canvas tutorial](/en-US/docs/Web/API/Canvas_API/Tutorial).

> ## Events and performance

Author: I've gone for "Optimizing event performance"

Events can be expensive for the browser to track and handle, especially when you are running an event continuously. For example, you might be tracking the position of the mouse using the [`mousemove`](/en-US/docs/Web/API/Element/mousemove_event) event to check whether it is still inside a certain area of the page:

```js
function handleMouseMove() {
  // Do stuff while mouse pointer is inside elem
}

elem.addEventListener("mousemove", handleMouseMove);
```

You might be running a `<canvas>` game in your page. While the mouse is inside the canvas, you will want to constantly check for mouse movement and cursor position and update the game state — including the score, the time, the position of all the sprites, collision detection information, etc. Once the game is over, you will no longer need to do all that, and in fact, it will be a waste of processing power to keeping listening for that event.

It is, therefore, a good idea to remove event listeners that are no longer needed. This can be done using {{domxref("EventTarget.removeEventListener", "removeEventListener()")}}:

```js
elem.removeEventListener("mousemove", handleMouseMove);
```

> Another tip is to use event delegation wherever possible. When you have some code to run in response to the user interacting with any one of a large number of child elements, you can set an event listener on their parent. Events fired on the child elements will bubble up to their parent, so you don't need to set the event listener on every child individually. Less event listeners to keep track of means better performance.

Reviewer: ```suggestion
Another tip is to use event delegation wherever possible. When you have some code to run in response to a user interacting with any one of the large number of child elements, you can set an event listener on their parent. Events fired on any child element will bubble up to their parent, so you don't need to set the event listener on each child individually. Less event listeners to keep track of means better performance.
```

Author: I mostly did this, but tweaked it a bit.

- **Reduce DOM manipulation**: Accessing and updating the DOM is computationally expensive, so you should minimize the amount that your JavaScript does, especially when performing constant DOM animation (see [Handling JavaScript animations](#handling_javascript_animations) above).
- **Batch DOM changes**: For essential DOM changes, you should batch them into groups that get done together, rather than just firing off each individual change as it occurs. This can reduce the amount of work the browser is doing in real terms, but also improve perceived performance. It can make the UI look smoother to get several updates out of the way in one go, rather than constantly making small updates. A useful tip here is — when you have a large chunk of HTML to add to the page, build the entire fragment first (typically inside a {{domxref("DocumentFragment")}}) and then append it all to the DOM in one go, rather than appending each item separately.
- **Simplify your HTML**: The simpler your DOM tree is, the faster it can be accessed and manipulated with JavaScript. Think carefully about what your UI needs, and remove unneccessary cruft.

> - **Reduce the amount of looped code**: Loops are expensive, so reduce the amount of loop usage in your code wherever possible. In cases where loops are unavoidable:

Reviewer: ```suggestion
- **Reduce the amount of looped code**: Loops are expensive, so reduce the amount of loop usage in your code wherever possible. In cases where loops are unavoidable, you can do a few things:
```

Author: I don't think this is needed, and "things" sounds a bit clumsy.

      </td>
    </tr>
    <tr>
      <th scope="row">Objectives:</th>
      <td>
        To learn about the effects of JavaScript on web performance,

>         and understand that JavaScript file size is not the only issue.

Author: I didn't like the whole bit about JavaScript file size, so I've rewritten it in a different way

      </td>
    </tr>
  </tbody>
</table>

## Optimizing JavaScript downloads

The most performant, least blocking JavaScript you can use is JavaScript that you don't use at all. You should use as little JavaScript as possible. Some tips to bear in mind:

- **You don't always need a framework**: You might be familiar with using a [JavaScript framework](/en-US/docs/Learn/Tools_and_testing/Client-side_JavaScript_frameworks). If you are experienced and confident with using this framework, and like all of the tooling it provides, then it might be your go-to tool for building most projects. However, frameworks are JavaScript-heavy. If you are creating a fairly static experience with few JavaScript requirements, you probably don't need that framework. You might be able to implement what you need using a few lines of standard JavaScript.

> - **Consider a simpler solution**: You might have a flashy, interesting idea about what solution to implement, but is it going to be appreciated by your users? Or would they prefer something simpler?

Author: updated

## Handling parsing and execution

Before looking at the tips contained in this section, it is important to talk about _where_ in the process of browser page rendering JavaScript is handled. When a web page is loaded:

1. The HTML is generally parsed first, in the order in which it appears on the page.
2. Whenever CSS is encountered, it is parsed to understand what styles need to be applied to the page. During this time linked assets such as images and web fonts start to be fetched.
3. Whenever JavaScript is encountered, it is parsed, evaluated, and run against the page.
4. Slightly later on, the browser works out how each HTML element should be styled, given the CSS applied to it.
5. The styled result is then painted to the screen.

> **Note:** This is a very simplified account of what happens, but it does give you an idea.

> The key step here is Step 3 — By default, JavaScript parsing and execution is render-blocking — it blocks parsing of the HTML that appears after the {{htmlelement("script")}} elements that refer to the JavaScript, and it therefore blocks styling and painting. As a result, you need to think carefully about not only what you are downloading, but when, and how that code is being executed.

Author: Ugh, my original paragraph was horrible. I've rewritten it in a slightly different way

First of all, you can add the `async` attribute to your `<script>` elements:

```html
<head>
  ...
  <script async src="main.js"></script>
  ...
</head>
```

This causes the script to be fetched in parallel with the DOM parsing, so it will be ready at the same time, and won't block rendering.

> **Note:** There is another attribute, `defer`, which causes the script to be executed after the document has been parsed, but before firing the [`DOMContentLoaded`](/en-US/docs/Web/API/Window/DOMContentLoaded_event) event. This has a similar effect to `async`.

You could also just not load the JavaScript at all until an event occurs when it is needed. This could be done via DOM scripting, for example:

```js
const scriptElem = document.createElement("script");
scriptElem.src = "index.js";
scriptElem.addEventListener("load", () => {
  // Run a function contained within index.js once it has definitely loaded
  init();
});
document.head.append(scriptElem);
```

JavaScript modules can be dynamically loaded using the {{jsxref("operators/import", "import()")}} function:

```js
import("./modules/myModule.js").then((module) => {
  // Do something with the module
});
```

> ### Breaking down long tasks

Author: Yes; updated!

      const task = tasks.shift();

      // Run the task
      task();
    }
  }
}
```

This allows you to avoid blocking the main thread when the user is actively interacting with the page, potentially providing a smoother user experience. However, by only yielding when necessary, we can continue running the current task when there are no user inputs to process. This also avoids tasks being placed at the back of the queue behind other non-essential browser-initiated tasks that were scheduled after the current one.

## Handling JavaScript animations

Animations can improve perceived performance, making interfaces feel snappier and making users feel like progress is being made when they are waiting for a page to load (loading spinners, for example). However, larger animations, and greater numbers of animations will naturally require more processing power to handle, which can degrade performance.

The most obvious piece of animation advice is to use less animations — cut out any non-essential animations, or consider giving your users a preference they can set to turn off animations, for example if they are using a low-powered device or a mobile device with limited battery power.

> For essential DOM animations, you are advised to use [CSS animations](/en-US/docs/Web/CSS/CSS_animations/Using_CSS_animations) where possible, rather than JavaScript animations (the [Web Animations API](/en-US/docs/Web/API/Web_Animations_API) provides a way to directly hook into CSS animations using JavaScript). Using the browser to directly perform DOM animations rather than manipulating inline styles using JavaScript is much faster and more efficient. See also [CSS performance optimization > Handling animations](/en-US/docs/Learn/Performance/CSS#handling_animations).

Author: Ah ha, no ;-)

The new CSS perf doc (content currently in https://github.com/mdn/content/pull/27558) has a section called "Handling animations", which is the one I'm referring to.

> For animations that can't be handled in JavaScript — for example animating an HTML {{htmlelement("canvas")}}, you are advised to use {{domxref("Window.requestAnimationFrame()")}} rather than older options such as {{domxref("setInterval()")}}. `requestAnimationFrame()` is specially designed for handling animation frames efficiently and consistently, for a smooth user experience. The basic pattern looks like this:

Reviewer: ```suggestion
For animations that can't be handled in JavaScript, for example, animating an HTML {{htmlelement("canvas")}}, you are advised to use {{domxref("Window.requestAnimationFrame()")}} rather than older options such as {{domxref("setInterval()")}}. The `requestAnimationFrame()` method is specially designed for handling animation frames efficiently and consistently, for a smooth user experience. The basic pattern looks like this:
```

```js
function loop() {
  // Clear the camvas before drawing the next frame of the animation
  ctx.fillStyle = "rgba(0, 0, 0, 0.25)";
  ctx.fillRect(0, 0, width, height);

  // Draw objects on the canvas and update their positioning data
  // ready for the next frame
  for (const ball of balls) {
    ball.draw();
    ball.update();
  }

  // Call requestAnimationFrame to run the loop() function again
  // at the right time to keep the animation smooth
  requestAnimationFrame(loop);
}

// Call the loop() function once to set the animation running
loop();
```

You can find a nice introduction to canvas animations at [Drawing graphics > Animations](/en-US/docs/Learn/JavaScript/Client-side_web_APIs/Drawing_graphics#animations), and a more in-depth example at [Object building practice](/en-US/docs/Learn/JavaScript/Objects/Object_building_practice). You can also find a full set of canvas tutorials at [Canvas tutorial](/en-US/docs/Web/API/Canvas_API/Tutorial).

> ### Events

Reviewer: is this supposed to be an H2? Also, could we add an action word here, such as "Optimizing events" or "Handling events"?

Author: Yes! Updated to H2, and updated heading to "Events and performance"

Events can be expensive for the browser to track and handle, especially when you are running an event continuously. For example, you might be tracking the position of the mouse using the [`mousemove`](/en-US/docs/Web/API/Element/mousemove_event) event to check whether it is still inside a certain area of the page:

```js
function handleMouseMove() {
  // Do stuff while mouse pointer is inside elem
}

elem.addEventListener("mousemove", handleMouseMove);
```

You might be running a `<canvas>` game in your page. While the mouse is inside the canvas, you will want to constantly check for mouse movement and cursor position and update the game state — including the score, the time, the position of all the sprites, collision detection information, etc. Once the game is over, you will no longer need to do all that, and in fact it will be a waste of processing power to keeping listening for that event.

> It is therefore a good idea to clean up after yourself and remove event listeners that are no longer needed. This can be done with {{domxref("EventTarget.removeEventListener", "removeEventListener()")}}:

Author: updated

```js
elem.removeEventListener("mousemove", handleMouseMove);
```

> Another tip is to use event delegation wherever possible — when you have some code to run when the user interacts with any one of a large number of child elements, you can set the event listener on their parent and have events that happen on them bubble up to their parent rather than having to set the event listener on every child individually. Less event listeners to handle means better performance.

Author: I've handled this in a slightly different way, but yeah, splitting it up into more sentences was needed.

> See [Event delegation](/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_delegation) for more details, and a useful example.

Reviewer: ```suggestion
See [Event delegation](/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_delegation) for more details and a useful example.
```

## Use more efficient code

There are several general best practices that will make your code run more efficiently.

> - **DOM manipulation is expensive**: Accessing and updating the DOM is computationally expensive, so you should minimize the amount that your JavaScript does, especially when performing constant DOM animation (see [Handling JavaScript animations](#handling_javascript_animations), above).

Reviewer: ```suggestion
- **Reduce DOM manipulation**: Accessing and updating the DOM is computationally expensive, so you should minimize the amount that your JavaScript does, especially when performing constant DOM animation (see [Handling JavaScript animations](#handling_javascript_animations) above).
```

- **Simplify your HTML**: The simpler your DOM tree is, the faster it can be accessed and manipulated with JavaScript. Think carefully about what your UI needs, and remove unneccessary cruft.
- **Use `getElementById()`**: {{domxref("Document.getElementById()")}} is more efficient for accessing DOM elements than other options such as {{domxref("Document.querySelector()")}}, despite the fact that the latter is more in-keeping with CSS best practices such as favouring classes over IDs because they have lower specificity. If your script involves a lot of DOM manipulation, consider using `getElementById()` if possible.
- **Avoid unnecessary variables**: Each variable your code uses adds a small amount of overhead, and these overheads can build up in a more complex script. For a start, manage your code carefully and remove unused variables. In addition, you should think about reducing your code down to remove variables. For example:

  ```js
  // This uses an unneccessary variable
  const pElem = document.getElementById("tagline");
  pElem.classList.add("fade-in");

  // This is more efficient
  document.getElementById("tagline").classList.add("fade-in");
  ```

> - **Loops are expensive**: If possible, reduce the amount of loop usage in your code. Where loops are unavoidable:

Author: Nope; the below items are things you should try to do where loops are totally unavoidable. I've rewritten it to make it clearer, using a lot of the text from your suggestion.

> - **Loops are expensive**: If possible, reduce the amount of loop usage in your code. Where loops are unavoidable:

Reviewer: Got it

- **Consider built-in browser features**: It might be that you can use a feature the browser already has, rather than creating your own via JavaScript. For example:
  - Use [built-in client-side form validation](/en-US/docs/Learn/Forms/Form_validation#using_built-in_form_validation).
  - Use the browser's own {{htmlelement("video")}} player.
  - Use [CSS animations](/en-US/docs/Web/CSS/CSS_animations/Using_CSS_animations) instead of a JavaScript animation library (see also [Handling animations](#handling_animations)).

You should also split your JavaScript into multiple files representing critical and non-critical parts. [JavaScript modules](/en-US/docs/Web/JavaScript/Guide/Modules) allow you to do this more efficiently than just using separate external JavaScript files. You split your code manually, but often a module bundler like [Webpack](https://webpack.js.org/) will do a better job of this.

> Then you can optimize these smaller files. [Minification](/en-US/docs/Glossary/Minification) reduces the number of characters in your file, thereby reducing the number of bytes or weight of your JavaScript. [Gzipping](/en-US/docs/Glossary/GZip_compression) compresses the file further and should be used even if you don't minify your code. [Brotli](/en-US/docs/Glossary/Brotli_compression) is similar to Gzip, but generally outperforms Gzip compression.

Author: Sounds good, but I've taken the coward's way out of fixing this for now by just moving the line about bundler tools to the end of the section. It now reads like this:

> You should also split your JavaScript into multiple files representing critical and non-critical parts. [JavaScript modules](/en-US/docs/Web/JavaScript/Guide/Modules) allow you to do this more efficiently than just using separate external JavaScript files.

> Then you can optimize these smaller files. [Minification](/en-US/docs/Glossary/Minification) reduces the number of characters in your file, thereby reducing the number of bytes or weight of your JavaScript. [Gzipping](/en-US/docs/Glossary/GZip_compression) compresses the file further and should be used even if you don't minify your code. [Brotli](/en-US/docs/Glossary/Brotli_compression) is similar to Gzip, but generally outperforms Gzip compression.

> You can split and optimize your code manually, but often a module bundler like [Webpack](https://webpack.js.org/) will do a better job of this.

- **Reduce DOM manipulation**: Accessing and updating the DOM is computationally expensive, so you should minimize the amount that your JavaScript does, especially when performing constant DOM animation (see [Handling JavaScript animations](#handling_javascript_animations) above).
- **Batch DOM changes**: For essential DOM changes, you should batch them into groups that get done together, rather than just firing off each individual change as it occurs. This can reduce the amount of work the browser is doing in real terms, but also improve perceived performance. It can make the UI look smoother to get several updates out of the way in one go, rather than constantly making small updates. A useful tip here is — when you have a large chunk of HTML to add to the page, build the entire fragment first (typically inside a {{domxref("DocumentFragment")}}) and then append it all to the DOM in one go, rather than appending each item separately.
- **Simplify your HTML**: The simpler your DOM tree is, the faster it can be accessed and manipulated with JavaScript. Think carefully about what your UI needs, and remove unneccessary cruft.

> - **Use `getElementById()`**: {{domxref("Document.getElementById()")}} is more efficient for accessing DOM elements than other options such as {{domxref("Document.querySelector()")}}, despite the fact that the latter is more in-keeping with CSS best practices such as favouring classes over IDs because they have lower specificity. If your script involves a lot of DOM manipulation, consider using `getElementById()` if possible.

Reviewer 2: I think it would take a truly absurd level of DOM manipulation for the differences between the two to be noticeable, I'd consider removing this one.

Author: OK, you got me on this one. I've removed it.

> - **Avoid unnecessary variables**: Each variable your code uses adds a small amount of overhead, and these overheads can build up in a more complex script. For a start, manage your code carefully and remove unused variables. In addition, you should think about reducing your code down to remove variables. For example:

Reviewer 2: Similarly, it would take a huge number of variables for this to make much difference, and it can really affect the readability of code - I reckon suggesting this is going to lead to a lot of premature optimisation.

Author: Removed this one too.

  // This is more efficient
  document.getElementById("tagline").classList.add("fade-in");
  ```

- **Reduce the amount of looped code**: Loops are expensive, so reduce the amount of loop usage in your code wherever possible. In cases where loops are unavoidable:

  - Avoid running the full loop when it is unneccessary, using {{jsxref("Statements/break", "break")}} or {{jsxref("Statements/continue", "continue")}} statements as appropriate. For example, if you are searching arrays for a specific name, you should break out of the loop once the name is found; there is no need to run further loop iterations:

    ```js
    function processGroup(array) {
      const toFind = "Bob";
      for (let i = 0; i < array.length - 1; i++) {
        if (array[i] === toFind) {
          processMatchingArray(array);
          break;
        }
      }
    }
    ```

  - Do work that is only needed once outside the loop. This may sound a bit obvious, but it is easy to overlook. As an example, take the above `for` loop:

    ```js
    // array.length is accessed on every iteration
    for(let i = 0; i < array.length - 1; i++) { ... };

    // Only access array.length once - more efficient

>     const arrLen = arraylength;

Author: I have committed the suggestion, but I totally take your point - the network request would be a much more realistic example. I'll update this once I've gone through and read your other comments.

>     const arrLen = arraylength;

Author: OK, better example now included in next commit ;-)
